app/geant-exporter/geant-exporter.cc

```diff
@@ -255,48 +109,153 @@ int to_pdg(const G4ProductionCutsIndex& index)
 
 //---------------------------------------------------------------------------//
 /*!
- * Write material table data to ROOT.
+ * Find the largest \c logical_volume.GetInstanceID() by recursively looping
+ * over all existing volumes. The \e InstanceID is the volume ID.
  *
- * The ROOT file must be open before this call.
+ * Function called by \c store_volumes(...) .
  */
-void store_geometry(TFile*                       root_file,
-                    const G4ProductionCutsTable& g4production_cuts_table,
-                    const G4VPhysicalVolume&     world_volume)
+void get_max_instance_id(G4LogicalVolume& logical_volume, int& max_instance_id)
 {
-    CELER_EXPECT(root_file);
+    max_instance_id = std::max(max_instance_id, logical_volume.GetInstanceID());
 
-    CELER_LOG(status) << "Exporting material and volume information";
+    // Recursively search for the max instance id
+    for (const auto i : celeritas::range(logical_volume.GetNoDaughters()))
+    {
+        get_max_instance_id(*logical_volume.GetDaughter(i)->GetLogicalVolume(),
+                            max_instance_id);
+    }
+}
 
-    TTree tree_materials("geometry", "geometry");
+//---------------------------------------------------------------------------//
+/*!
+ * Store all logical volumes by recursively looping over them.
+ *
+ * Function called by \c store_volumes(...) .
+ */
+void loop_volumes(std::vector<ImportVolume>& volumes,
+                  const G4LogicalVolume&     logical_volume)
+{
+    // Add volume to vector
+    ImportVolume volume;
+    volume.material_id = logical_volume.GetMaterialCutsCouple()->GetIndex();
+    volume.name        = logical_volume.GetName();
+    volume.solid_name  = logical_volume.GetSolid()->GetName();
 
-    // Create geometry map and ROOT branch
-    GdmlGeometryMap geometry;
-    TBranch* branch = tree_materials.Branch("GdmlGeometryMap", &geometry);
-    CELER_ASSERT(branch);
+    // Volume position represents the volume id
+    volumes[logical_volume.GetInstanceID()] = volume;
+
+    // Recursive: repeat for every daughter volume, if there are any
+    for (const auto i : celeritas::range(logical_volume.GetNoDaughters()))
+    {
+        loop_volumes(volumes,
+                     *logical_volume.GetDaughter(i)->GetLogicalVolume());
+    }
```

Comment:
This loop looks dangerous from a scaling perspective: every volume loops recursively into all its daughters, so that means this could be an `O(N^2)` operation. Is it possible to loop directly over logical volume IDs rather than hierarchically through the geometry? That would allow you to eliminate the `get_max_instance_id` function and juts directly loop over volumes. If not, it might be better here to temporarily use a map (or unordered map) of volume IDs -> volumes, and then convert to a vector by looping over indices.

Reply:
Yes. This loop has been debated between @mrguilima and me. It is a very inefficient loop that I am still not sure how to tackle, and I've been playing around with it since I pushed these commits. The `loop_volumes` ends up going over all geometry nodes (which represents a total of 2,104,795 nodes for the case of cms2018.gdml), because there is no way to loop over instances, as you said. There is no `GetInstanceID(G4int id)` or `GetLogicalVolume(G4int instance_id)`, so the only obvious way was to deal with it the hard way.

Reply:
Thumbs down for the code, not for you :) you still might be able to recurse only once if you use a map rather than a vector

Reply:
At the end, looping over all nodes seems to be inevitable (I was hoping that some loops over daughter volumes could be skipped). So I accepted defeat and went for the map solution you mentioned. At least we loop only once over them to build the map and use it to populate the final `vector<ImportVolume>`.